File: Java/src/main/java/baekJoon/shortestPath/FloydWarshall/floyd2_11780.java
package baekJoon.shortestPath.FloydWarshall;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.*;

import static Util.Constants.INPUT;

// 플로이드 와샬 알고리즘 응용 문제 => 각 점
public class floyd2_11780 {
    static int N, M;
    static int[][] d;
    static int INF = Integer.MAX_VALUE;
    static List<Integer>[][] set;
    public static void main(String[] args) throws Exception{
        System.setIn(new FileInputStream(INPUT));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        M = Integer.parseInt(br.readLine());
        d = new int[N + 1][N + 1];
        set = new List[N + 1][N + 1];
        for(int i = 1; i <= N; i++) {
            for(int j = 1; j <= N; j++) {
                d[i][j] = INF;
                set[i][j] = new ArrayList<>();
            }
            d[i][i] = 0;
        }
        for(int i = 0; i < M; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            d[a][b] = Math.min(d[a][b], c);
        }
        for(int k = 1; k <= N; k++) {
            for(int i = 1; i <= N; i++) {
                for(int j = 1; j <= N; j++) {
                    if(i == j) continue;
                    if(d[i][k] != INF && d[k][j] != INF) {
                        if(d[i][j] > d[i][k] + d[k][j]) {
                            d[i][j] = d[i][k] + d[k][j];
                            if(!set[i][j].isEmpty()) set[i][j].clear();
                            if(!set[i][k].isEmpty()) {
                                set[i][j].addAll(set[i][k]);
                            }
                            set[i][j].add(k);
                            if(!set[k][j].isEmpty()) {
                                set[i][j].addAll(set[k][j]);
                            }
                        }
                    }
                }
            }
        }
        for(int i = 1; i <= N; i++) {
            for(int j = 1; j <= N; j++) {
                if(d[i][j] == INF) System.out.print(0 + " ");
                else System.out.print(d[i][j] + " ");
            }
            System.out.println();
        }
        for(int k = 1 + N; k <= N * N + N; k++) {
            int i = k / N;
            int j = k % N;
            if(j == 0) {
                i--;
                j += N;
            }
            if(d[i][j] == INF) System.out.println(0);
            else {
                if(set[i][j].isEmpty()) {
                    if(i == j) System.out.println(0);
                    else {
                        System.out.println(2 + " " + i + " " + j);
                    }
                }
                else {
                    System.out.print(2 + set[i][j].size() + " " + i + " ");
                    for(int val : set[i][j]) {
                        System.out.print(val + " ");
                    }
                    System.out.print(j);
                    System.out.println();
                }
            }
        }
    }
}

File: Kotlin/src/baekJoon/ShortestPath/floyd_warshall/Floyd2_11780.kt
package baekJoon.ShortestPath.floyd_warshall

import INPUT
import printMatrix
import java.io.FileInputStream
import java.lang.StringBuilder
import kotlin.math.min

class Floyd2_11780 {

    private var n = 0
    private var m = 0
    private lateinit var dp: Array<Array<Int>>
    private lateinit var path: Array<Array<ArrayList<Int>>>
    private val INF = 987654321

    fun solution() {
        System.setIn(FileInputStream(INPUT))
        val br = System.`in`.bufferedReader()
        n = br.readLine().toInt()
        m = br.readLine().toInt()
        dp = Array(n + 1) { Array(n + 1) { INF } }
        for(i in 1..n) {
            for(j in 1..n) {
                if(i == j) dp[i][j] = 0
            }
        }
        path = Array(n + 1) { Array(n + 1) { arrayListOf() } }
        repeat(m) {
            val (a, b, c) = br.readLine().split(" ").map { it.toInt() }
            dp[a][b] = min(dp[a][b], c) // 같은 경로의 값이 입력될 수 있으므로
        }
        for(k in 1..n) {
            for(i in 1..n) {
                for(j in 1..n) {
                    if(i == j) continue
                    if(dp[i][j] > dp[i][k] + dp[k][j]) {
                        dp[i][j] = dp[i][k] + dp[k][j]
                        path[i][j].clear()
                        path[i][j].addAll(path[i][k])
                        path[i][j].add(k)
                        path[i][j].addAll(path[k][j])
                    }
                }
            }
        }
        for(i in 1..n) {
            for(j in 1..n) {
                print("${if(dp[i][j] == INF) 0 else dp[i][j]} ")
            }
            println()
        }
        for(k in (1 * n)until(n * n) + n) {
            val i = k / n
            val j = k % n + 1
            if(i == j) println(0)
            else {
                if(path[i][j].isEmpty()) {
                    println("2 $i $j")
                } else {
                    val sb = StringBuilder("${2 + path[i][j].size} $i ")
                    for(v in path[i][j]) {
                        sb.append("$v ")
                    }
                    sb.append("$j")
                    println(sb.toString())
                }
            }
        }
    }
}

fun main() {
    Floyd2_11780().solution()
}